File: Python/Sqrt(x).py
"""
Implement int sqrt(int x).
Compute and return the square root of x.
"""

class Solution1:
    # @param x, an integer
    # @return an integer
    def sqrt(self, x):
        start = 0
        end = x
        
        while start < end:
            mid = (int)((start+end)/2)
            if mid*mid == x:
                return mid
            elif mid*mid > x:
                end = mid-1
            else:
                start = mid+1
        if start * start > x:
            return start -1
        else:
            return start

# Newton's method, works for double input
class Solution2(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        r = x
        while r*r > x:
            r = (r + x/r) / 2
        return r

File: Java/Sqrt(x).java
/*
Implement int sqrt(int x).

Compute and return the square root of x.
*/

public class Solution {
    public int mySqrt(int x) {
        int lo = 1;
        int hi = x / 2;
        
        if(x < 2) return x;

        while (hi >= lo) {     
            int mid = lo+(hi-lo)/2;
            if (x / mid < mid) {
                hi = mid-1;
            } else {
                lo = mid+1;  
            }
        }
        return hi;
    }
}
